third_party/nvidia/lib/TritonNVIDIAGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -24,60 +26,77 @@ using ::mlir::triton::gpu::SharedEncodingAttr;
-    assert(ll.has_value() && "Failed to convert layout to linear layout");
-    auto freeVariableMasks = ll->getFreeVariableMasks();
-    auto regMasks = freeVariableMasks[kReg];
-    if (regMasks & regIdx) {
-      // Step 1: check register redundancy
-      mask = false_val();
-    } else {
-      Value warpSize =
-          i32_val(triton::gpu::TritonGPUDialect::getThreadsPerWarp(moduleOp));
-      Value laneId = urem(tid, warpSize);
-      Value warpId = udiv(tid, warpSize);
-      // Step 2: check lane and warp redundancy
-      auto laneMasks = freeVariableMasks[kLane];
-      auto warpMasks = freeVariableMasks[kWarp];
-      mask = and_(mask, icmp_eq(and_(i32_val(laneMasks), laneId), i32_val(0)));
-      mask = and_(mask, icmp_eq(and_(i32_val(warpMasks), warpId), i32_val(0)));
-      if (numCTAs > 1) {
-        // Step 3: check block redundancy
-        auto ctaId = targetInfo.getClusterCTAId(rewriter, loc);
-        auto ctaMasks = freeVariableMasks[kBlock];
-        mask = and_(mask, icmp_eq(and_(i32_val(ctaMasks), ctaId), i32_val(0)));
-      }
-    }
-  } else {
-    mask = and_(mask, icmp_eq(tid, i32_val(0)));
-    if (numCTAs > 1) {
-      auto ctaId = targetInfo.getClusterCTAId(rewriter, loc);
-      // If the tensor is not ranked, then it is a scalar and only thread 0 of
-      // CTA0 within the cluster can write
-      mask = and_(mask, icmp_eq(ctaId, i32_val(0)));
+
+  int32_t fullMask = -1;
+  llvm::MapVector<StringAttr, int32_t> ret;
+  for (auto dimName : {kReg, kLane, kWarp, kBlock}) {
+    ret[dimName] = fullMask;
+  }
+  return ret;
+}
+
+llvm::MapVector<StringAttr, int32_t> getFreeVariableMasks(Type type) {
+  auto ctx = type.getContext();
+  auto tensorTy = dyn_cast<RankedTensorType>(type);
+  if (!tensorTy) {
+    return getAllFreeVarMasks(ctx);
+  }
+
+  auto ll = ttg::toLinearLayout(tensorTy.getShape(), tensorTy.getEncoding());
+  assert(ll && "failed to convert to linear layout");
+  return ll->getFreeVariableMasks();
+}
+
+Value maybeAnd(RewriterBase &rewriter, Location loc, Value a, Value b) {
+  if (a && b) {
+    return and_(a, b);
+  }
+  return a ? a : b;
+}
+
+// Return a predicate that is true only if the current thread holds unique data,
+// according to freeVarsMask. The predicate may be null to indicate no
+// predication is required.
+Value emitRedundantThreadPredicate(
+    ModuleOp moduleOp, const llvm::MapVector<StringAttr, int32_t> &freeVarMasks,
+    ConversionPatternRewriter &rewriter, Location loc,
+    const NVIDIA::TargetInfo &targetInfo) {
+  auto ctx = rewriter.getContext();
+  auto kLane = str_attr("lane");
+  auto kWarp = str_attr("warp");
+  auto kBlock = str_attr("block");
+
+  auto warpSize = triton::gpu::TritonGPUDialect::getThreadsPerWarp(moduleOp);
+  auto emitBlockId = freeVarMasks.lookup(kBlock) != 0;
+  auto [laneId, warpId, blockId] =
+      emitHardwareTuple(loc, rewriter, targetInfo, emitBlockId, warpSize);
+  auto zero = i32_val(0);
+
+  Value pred;
```

Comment:
Hi @peterbell10! Should it be like that?

`Value pred = b.true_val();`

Reply:
This is intentional, as I wrote in the description
> I also return null values if the predicate is always true, which allows us to omit the predicate entirely if there is no redundant data.

Reply:
Thanks for the clarification!